x/gov/keeper.go

```diff
@@ -0,0 +1,415 @@
+	vote := Vote{
+		ProposalID: proposalID,
+		Voter:      voter,
+		Option:     option,
+	}
+	keeper.setVote(ctx, proposalID, voter, vote)
+
+	return nil
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) GetVote(ctx sdk.Context, proposalID int64, voter sdk.Address) *Vote {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := store.Get([]byte(fmt.Sprintf("%d", proposalID) + ":votes:" + fmt.Sprintf("%s", voter)))
+	if bz == nil {
+		return nil
+	}
+	vote := &Vote{}
+	keeper.cdc.MustUnmarshalBinary(bz, vote)
+	return vote
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) setVote(ctx sdk.Context, proposalID int64, voter sdk.Address, vote Vote) {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := keeper.cdc.MustMarshalBinary(vote)
+	key := []byte(fmt.Sprintf("%d", proposalID) + ":votes:" + fmt.Sprintf("%s", voter))
+	store.Set(key, bz)
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) GetVotes(ctx sdk.Context, proposalID int64) sdk.Iterator {
+	store := ctx.KVStore(keeper.storeKey)
+	return sdk.KVStorePrefixIterator(store, []byte(fmt.Sprintf("%d", proposalID)+":votes:"))
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) deleteVote(ctx sdk.Context, proposalID int64, voter sdk.Address) {
+	store := ctx.KVStore(keeper.storeKey)
+	key := []byte(fmt.Sprintf("%d", proposalID) + ":votes:" + fmt.Sprintf("%s", voter))
+	store.Delete(key)
+}
+
+// =====================================================
+// Deposits
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) GetDeposit(ctx sdk.Context, proposalID int64, depositer sdk.Address) *Deposit {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := store.Get([]byte(fmt.Sprintf("%d", proposalID) + ":deposits:" + fmt.Sprintf("%s", depositer)))
+	if bz == nil {
+		return nil
+	}
+	deposit := &Deposit{}
+	keeper.cdc.MustUnmarshalBinary(bz, deposit)
+	return deposit
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) setDeposit(ctx sdk.Context, proposalID int64, depositer sdk.Address, deposit Deposit) {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := keeper.cdc.MustMarshalBinary(deposit)
+	key := []byte(fmt.Sprintf("%d", proposalID) + ":deposits:" + fmt.Sprintf("%s", depositer))
+	store.Set(key, bz)
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) AddDeposit(ctx sdk.Context, proposalID int64, depositer sdk.Address, depositAmount sdk.Coins) sdk.Error {
+	proposal := keeper.GetProposal(ctx, proposalID)
+	if proposal == nil {
+		return ErrUnknownProposal(proposalID)
+	}
+
+	if (proposal.Status != "DepositPeriod") && (proposal.Status != "VotingPeriod") {
+		return ErrAlreadyFinishedProposal(proposalID)
+	}
+
+	_, _, err := keeper.ck.SubtractCoins(ctx, depositer, depositAmount)
+	if err != nil {
+		return err
+	}
+
+	proposal.TotalDeposit = proposal.TotalDeposit.Plus(depositAmount)
```

Comment:
we're subtracting coins and adding them to a non-banker object here - we should create an invariance test for when these are returned (are they?) - I imagine no coins are destroyed?

Reply:
The keeper_test does have an invariance test.